tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,260 @@
+      }
+      return tf.withSubScope("staticDimsCheckSuccess")
+          .withControlDependencies(Collections.EMPTY_LIST)
+          .noOp();
+    }
+    // Dynamic checks.
+    Operand<TInt32> weightsShape = tf.shape(sampleWeights);
+    Operand<TInt32> weightsRank = tf.rank(sampleWeights);
+    Operand<TInt32> valuesShape = tf.shape(values);
+    Operand<TInt32> valuesRank = tf.rank(values);
+
+    Operand<TBool> isScalar = tf.math.equal(weightsRank, tf.constant(0));
+    List<Operand<?>> data =
+        Arrays.asList(
+            tf.constant(ASSERT_BROADCAST_ERROR_PREFIX),
+            tf.constant("weights.shape="),
+            weightsShape,
+            tf.constant("values.shape="),
+            valuesShape,
+            tf.constant("isScalar="),
+            isScalar);
+
+    Operand<TBool> validNonsclar =
+        canBroadcastNonscalarShapes(tf, weightsRank, weightsShape, valuesRank, valuesShape);
+
+    Operand<TBool> isValidShape = tf.select(isScalar, isScalar, validNonsclar);
+
+    return tf.withSubScope("broadcastWeights-dynamic").assertThat(isValidShape, data);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have the same rank and valid dimensions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsRank the operand for the rank of the sample weights
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesRank the operand for the rank of the sample weights
+   * @param valuesShape the operand for the shape of the sample weights
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the Shape is scalar or not.
+   */
+  private static <T extends TNumber> Operand<TBool> canBroadcastNonscalarShapes(
+      Ops tf,
+      Operand<T> weightsRank,
+      Operand<T> weightsShape,
+      Operand<T> valuesRank,
+      Operand<T> valuesShape) {
+    tf = tf.withSubScope("hasValidNonscalarShape");
+    Operand<TBool> isSameRank = tf.math.equal(valuesRank, weightsRank);
+    return tf.select(isSameRank, canBroadcastDims(tf, weightsShape, valuesShape), isSameRank);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have valid dimensions or not.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesShape the operand for the shape of the values
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the shapes have valid dimensions or not.
+   */
+  private static <T extends TNumber> Operand<TBool> canBroadcastDims(
+      Ops tf, Operand<T> weightsShape, Operand<T> valuesShape) {
+    tf = tf.withSubScope("hasValidDims");
+    Operand<T> valuesShape2d = tf.expandDims(valuesShape, tf.constant(-1));
+    Operand<T> validDims =
+        tf.concat(Arrays.asList(valuesShape2d, tf.onesLike(valuesShape2d)), tf.constant(1));
+    SetDiff1d<T, TInt32> invalidDimsDiff = tf.setDiff1d(weightsShape, tf.shape.flatten(validDims));
```

Comment:
The problem is there is no `set_difference` equivalent in TensorFlow Java. The is only `setDiff1d`.

Reply:
We may be able to use DenseToDenseSetOperation.java. Let me investigate it.

Reply:
Would it work to evaluate the element-wise counterpart of (weightsShape = 1 or weightsShape = valuesShape)?

Reply:
I traced the python code for `set_difference` and it ends up calling a C++ routine for `_set_operation` with a set operation of `a-b`. The c++ routine ends up calling one of the sparse operations, `denseToDenseSetOperation` or `denseToSparseSetOperation`.  Because we are not supporting Sparse Tensors at the moment, that leaves `denseToDenseSetOperation`. I have it coded and it seems to work, but I am getting a shape mismatch error on `testValidScalar`, that I am now tracking down. The other tests in `AssertBroadcastableTest` are passing.

Reply:
I created a new class `SetOps` that implements the difference, intersection and union set operations just as `TF/python/ps/sets_impl.py` does. Currently, these utilize the `tf.sparse.denseToDenseSetOperation` method. I also created UnitTests for these.  
I did have to do a hack on scalar values in `MetricsHelper`  to work around the non-lazy `select` for `isValidShape`, otherwise `validNonscalar` fails on a scalar weight shape being `0`. If `select` were lazy, the `validNonscalar`  branch wouldn't get executed when `isScalar` is true.

Reply:
I raised concerns in a comment on the `SetOps` implementation. If I'm right about the problems I saw there, then I again find myself wondering if it would work more simply to evaluate the element-wise counterpart of (weightsShape = 1 or weightsShape = valuesShape)?